## Aligning reads to a genome
Once the reads look good, then we need to align them to a genome so that we can make sense of them. For this, we need (1) a program to use, (2) the reads we want to align, and (3) the genome that we want to align the reads to. The common alignment programs that are used for Illumina short reads include [bwa](https://bio-bwa.sourceforge.net/bwa.shtml) and [bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml). These are both available as modules: `module load bwa` // `module load bowtie2`.

We also need a reference genome to use! Look on NCBI genome database if you don’t have one → can download the genome file (for today, we’ll be using the Arctic charr reference genome: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_002910315.2/, which I’ve already downloaded to our shared directory) 
I’ve already indexed the genome, but to do so with your own genome, you’ll need to use bwa index genome.fa.gz and bowtie2-build genome.fa.gz (this can take a while, depending on the size of your genome)

Other aligners may be better for your purposes in the future, depending on what kind of data you’re working with. There also are “workflow” programs that can be used to do many of these steps together: [STACKS](https://catchenlab.life.illinois.edu/stacks/manual/), [ipyrad](https://ipyrad.readthedocs.io/en/master/index.html), etc. (but these usually use bwa or bowtie2 under the hood).
